File: tam_agent/report/formatter.py
"""Console output formatting using Rich."""

from __future__ import annotations

from rich.console import Console
from rich.panel import Panel
from rich.table import Table

from tam_agent.models import FinalReport

console = Console()


def print_report(report: FinalReport) -> None:
    """Print a formatted report to the console."""
    r = report
    c = r.company

    # Header
    console.print()
    console.print(
        Panel(
            f"[bold]{c.name}[/bold]\n"
            f"{c.industry} | {c.geography}\n"
            f"Products: {', '.join(c.products_services)}",
            title="[bold green]TAM/SOM Assessment Report[/bold green]",
            border_style="green",
        )
    )

    # Industry Classification
    console.print(f"\n[bold]Industry Classification[/bold]")
    console.print(
        f"  NAICS {r.industry.naics_code}: {r.industry.naics_description} "
        f"(confidence: {r.industry.confidence:.0%})"
    )
    console.print(f"  Sub-segments: {', '.join(r.industry.sub_segments)}")

    # TAM Table
    tam_table = Table(title="\nTotal Addressable Market (TAM)", show_lines=True)
    tam_table.add_column("Method", style="cyan")
    tam_table.add_column("Estimate", style="green", justify="right")
    tam_table.add_column("Year", justify="center")
    tam_table.add_column("Confidence", justify="center")

    tam_table.add_row(
        "Top-Down",
        f"${r.tam_top_down.value_usd:,.0f}",
        str(r.tam_top_down.year),
        f"{r.tam_top_down.confidence:.0%}",
    )
    tam_table.add_row(
        "Bottom-Up",
        f"${r.tam_bottom_up.value_usd:,.0f}",
        str(r.tam_bottom_up.year),
        f"{r.tam_bottom_up.confidence:.0%}",
    )
    tam_table.add_row(
        "[bold]Consensus[/bold]",
        f"[bold]${r.tam_consensus.value_usd:,.0f}[/bold]",
        str(r.tam_consensus.year),
        f"[bold]{r.tam_consensus.confidence:.0%}[/bold]",
    )
    console.print(tam_table)

    # Competitive Landscape
    comp_table = Table(title="\nCompetitive Landscape", show_lines=True)
    comp_table.add_column("Competitor", style="cyan")
    comp_table.add_column("Est. Revenue", justify="right")
    comp_table.add_column("Market Share", justify="center")

    for comp in r.competitive_landscape.top_competitors[:8]:
        rev = f"${comp.estimated_revenue_usd:,.0f}" if comp.estimated_revenue_usd else "N/A"
        share = f"{comp.market_share_pct:.1f}%" if comp.market_share_pct else "N/A"
        comp_table.add_row(comp.name, rev, share)

    console.print(comp_table)
    console.print(
        f"  Market concentration: [bold]{r.competitive_landscape.market_concentration}[/bold] "
        f"| ~{r.competitive_landscape.total_competitors_estimated} total competitors"
    )

    # SOM
    console.print(
        Panel(
            f"[bold green]${r.som.value_usd:,.0f}[/bold green]\n"
            f"Market Share: {r.som.market_share_pct:.1f}%  |  "
            f"Confidence: {r.som.confidence:.0%}\n\n"
            f"{r.som.methodology}",
            title="[bold]Serviceable Obtainable Market (SOM)[/bold]",
            border_style="yellow",
        )
    )

    # Methodology
    console.print(f"\n[bold]Methodology Notes[/bold]")
    console.print(f"  {r.methodology_notes[:500]}{'...' if len(r.methodology_notes) > 500 else ''}")
    console.print()

File: src/report/formatter.ts
/** Console output formatting. */

import type { FinalReport } from "../models.js";
import { formatUsd, formatPct } from "../util/format.js";

export function printReport(report: FinalReport): void {
  const r = report;
  const c = r.company;

  const divider = "═".repeat(60);
  const thinDiv = "─".repeat(60);

  // Header
  console.log(`\n\x1b[32m${divider}\x1b[0m`);
  console.log(`\x1b[1;32m  TAM/SOM Assessment Report\x1b[0m`);
  console.log(`\x1b[32m${divider}\x1b[0m`);
  console.log(`  \x1b[1m${c.name}\x1b[0m`);
  console.log(`  ${c.industry} | ${c.geography}`);
  console.log(`  Products: ${c.products_services.join(", ")}`);

  // Industry Classification
  console.log(`\n\x1b[1mIndustry Classification\x1b[0m`);
  console.log(
    `  NAICS ${r.industry.naics_code}: ${r.industry.naics_description} ` +
      `(confidence: ${formatPct(r.industry.confidence)})`,
  );
  console.log(`  Sub-segments: ${r.industry.sub_segments.join(", ")}`);

  // TAM Table
  console.log(`\n\x1b[1mTotal Addressable Market (TAM)\x1b[0m`);
  console.log(`  ${thinDiv}`);
  console.log(
    `  ${"Method".padEnd(15)} ${"Estimate".padStart(20)} ${"Year".padStart(6)} ${"Confidence".padStart(12)}`,
  );
  console.log(`  ${thinDiv}`);
  console.log(
    `  ${"Top-Down".padEnd(15)} ${formatUsd(r.tam_top_down.value_usd).padStart(20)} ${String(r.tam_top_down.year).padStart(6)} ${formatPct(r.tam_top_down.confidence).padStart(12)}`,
  );
  console.log(
    `  ${"Bottom-Up".padEnd(15)} ${formatUsd(r.tam_bottom_up.value_usd).padStart(20)} ${String(r.tam_bottom_up.year).padStart(6)} ${formatPct(r.tam_bottom_up.confidence).padStart(12)}`,
  );
  console.log(
    `  \x1b[1m${"Consensus".padEnd(15)} ${formatUsd(r.tam_consensus.value_usd).padStart(20)} ${String(r.tam_consensus.year).padStart(6)} ${formatPct(r.tam_consensus.confidence).padStart(12)}\x1b[0m`,
  );
  console.log(`  ${thinDiv}`);

  // Competitors
  console.log(`\n\x1b[1mCompetitive Landscape\x1b[0m`);
  console.log(`  ${thinDiv}`);
  console.log(
    `  ${"Competitor".padEnd(25)} ${"Est. Revenue".padStart(18)} ${"Market Share".padStart(14)}`,
  );
  console.log(`  ${thinDiv}`);
  for (const comp of r.competitive_landscape.top_competitors.slice(0, 8)) {
    const rev = comp.estimated_revenue_usd
      ? formatUsd(comp.estimated_revenue_usd)
      : "N/A";
    const share =
      comp.market_share_pct != null
        ? `${comp.market_share_pct.toFixed(1)}%`
        : "N/A";
    console.log(
      `  ${comp.name.padEnd(25)} ${rev.padStart(18)} ${share.padStart(14)}`,
    );
  }
  console.log(`  ${thinDiv}`);
  console.log(
    `  Market concentration: \x1b[1m${r.competitive_landscape.market_concentration}\x1b[0m ` +
      `| ~${r.competitive_landscape.total_competitors_estimated} total competitors`,
  );

  // SOM
  console.log(`\n\x1b[33m${divider}\x1b[0m`);
  console.log(`\x1b[1m  Serviceable Obtainable Market (SOM)\x1b[0m`);
  console.log(`\x1b[33m${divider}\x1b[0m`);
  console.log(`  \x1b[1;32m${formatUsd(r.som.value_usd)}\x1b[0m`);
  console.log(
    `  Market Share: ${r.som.market_share_pct.toFixed(1)}%  |  ` +
      `Confidence: ${formatPct(r.som.confidence)}`,
  );
  console.log(`  ${r.som.methodology}`);

  // Methodology
  console.log(`\n\x1b[1mMethodology Notes\x1b[0m`);
  const notes = r.methodology_notes;
  console.log(`  ${notes.length > 500 ? notes.slice(0, 500) + "..." : notes}`);
  console.log();
}
